### PR TITLE
Drools 7383 full reliable strategy - basic tests

### PR DESCRIPTION
**Pull request by nprentza**
The branch includes changes suggested in patches 
https://github.com/mariofusco/drools/commit/8f78bff36da1059ad3e533ed830c58b561e9a373.patch, and, 
https://gist.github.com/mariofusco/d274a8541913d70bee8ee6c8efe3b46e

and other changes to  support drools-reliability FULL strategy. 

It also includes basic methods for testing FULL strategy: 
- insertFailover_propListShouldNotBeEmpty
- insertFireFailover_shouldNotRepeatFiredMatch
- insertUpdateFailover_shouldNotFiredMatch
- insertNonMatching_Failover_UpdateWithMatching_ShouldFiredMatch

### CODE REVIEW

**Comment by kie-ci4** [comment]
**(tests) - serverless-workflow-examples job** `#587` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>


export BUILD_MVN_OPTS_CURRENT= 
build-chain build cross_pr -f 'https://raw.githubusercontent.com/kiegroup/drools/main/.ci/buildchain-config.yaml' -o 'bc' -p kiegroup/kogito-examples -u https://github.com/kiegroup/drools/pull/5239 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/drools/job/main/job/pullrequest/job/drools.tests.downstream.serverless-workflow-examples/587/display/redirect

**Test results:**
- PASSED: 79
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/drools/job/main/job/pullrequest/job/drools.tests.downstream.serverless-workflow-examples/587/testReport/org.kie.kogito.examples/DataIndexRestIT/testDataIndexRest/">org.kie.kogito.examples.DataIndexRestIT.testDataIndexRest</a></summary>
1 expectation failed.<br/>JSON path data.ProcessInstances[0].state doesn't match.<br/>Expected: is "COMPLETED"<br/>  Actual: ACTIVE<br/>
</details>

File: drools-core/src/main/java/org/drools/core/phreak/PropagationEntry.java [hunk 1]
```diff
@@ -285,12 +294,30 @@ public String toString() {
         public InternalFactHandle getHandle() {
             return handle;
         }
+
+        @Override
+        public void writeExternal(ObjectOutput out) throws IOException {
+            out.writeObject(next);
+            out.writeObject(handle);
+            out.writeObject(context);
+            //out.writeObject(objectTypeConf);
```
Comment (mariofusco) [inline drools-core/src/main/java/org/drools/core/phreak/PropagationEntry.java]: Please remove this comment out line.

File: drools-core/src/main/java/org/drools/core/phreak/PropagationEntry.java [hunk 2]
```diff
@@ -285,12 +294,30 @@ public String toString() {
...
+        }
+
+        @Override
+        public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
+            this.next = (PropagationEntry) in.readObject();
+            this.handle = (InternalFactHandle) in.readObject();
+            this.context = (PropagationContext) in.readObject();
+            //this.objectTypeConf = (ObjectTypeConf) in.readObject();
```
Comment (mariofusco) [inline drools-core/src/main/java/org/drools/core/phreak/PropagationEntry.java]: Same as above.

File: drools-core/src/main/java/org/drools/core/phreak/PropagationEntry.java [hunk 3]
```diff
@@ -221,6 +226,10 @@ public static void execute( InternalFactHandle handle, PropagationContext contex
         }
 
         private static void propagate( InternalFactHandle handle, PropagationContext context, ReteEvaluator reteEvaluator, ObjectTypeConf objectTypeConf ) {
+            if (objectTypeConf == null) {
+                // it can be null after deserialization
+                objectTypeConf = handle.getEntryPoint(reteEvaluator).getObjectTypeConfigurationRegistry().getOrCreateObjectTypeConf(handle.getEntryPointId(), handle.getObject());
```
Comment (mariofusco) [inline drools-core/src/main/java/org/drools/core/phreak/PropagationEntry.java:231]: This line could be encapsulated in a `getOrCreateObjectTypeConf` method on the `InternalFactHandle`. Just a reminder, leave it as it is for now.

File: drools-core/src/main/java/org/drools/core/phreak/SynchronizedPropagationList.java [hunk 4]
```diff
@@ -15,31 +15,35 @@
...
+    protected volatile boolean hasEntriesDeferringExpiration = false;
 
-    private volatile boolean firingUntilHalt = false;
+    protected volatile boolean firingUntilHalt = false;
 
     public SynchronizedPropagationList(ReteEvaluator reteEvaluator) {
         this.reteEvaluator = reteEvaluator;
     }
 
+    public SynchronizedPropagationList(){
```
Comment (mariofusco) [inline drools-core/src/main/java/org/drools/core/phreak/SynchronizedPropagationList.java]: I guess this empty constructor is only to support serialization, correct?

File: drools-core/src/main/java/org/drools/core/phreak/SynchronizedPropagationList.java [hunk 5]
```diff
@@ -15,31 +15,35 @@
...
+        //this.reteEvaluator=null;
```
Comment (mariofusco) [inline drools-core/src/main/java/org/drools/core/phreak/SynchronizedPropagationList.java]: Please remove this comment out line.

File: drools-reliability/drools-reliability-core/src/main/java/org/drools/reliability/core/ReliableAgenda.java [hunk 6]
```diff
@@ -46,7 +46,8 @@ protected PropagationList createPropagationList() {
             propagationList = new ReliablePropagationList(workingMemory);
             componentsStorage.put("PropagationList", propagationList);
         } else {
-            propagationList = new ReliablePropagationList(workingMemory, propagationList);
+            //propagationList = new ReliablePropagationList(workingMemory, propagationList);
```
Comment (mariofusco) [inline drools-reliability/drools-reliability-core/src/main/java/org/drools/reliability/core/ReliableAgenda.java]: Please remove this comment out line.

File: drools-reliability/drools-reliability-core/src/main/java/org/drools/reliability/core/ReliableSessionInitializer.java [hunk 7]
```diff
@@ -107,7 +108,46 @@ static class FullReliableSessionInitializer implements SessionInitializer {
 
         @Override
         public InternalWorkingMemory init(InternalWorkingMemory session, PersistedSessionOption persistedSessionOption) {
+            if (!persistedSessionOption.isNewSession()) {
+                // re-propagate objects from the storage to the new session
+                populateSessionFromStorage(session);
+            }
+            //session.setWorkingMemoryActionListener(entry -> onWorkingMemoryAction(session, entry));
```
Comment (mariofusco) [inline drools-reliability/drools-reliability-core/src/main/java/org/drools/reliability/core/ReliableSessionInitializer.java]: Please remove this comment out line.

File: drools-reliability/drools-reliability-infinispan/src/test/java/org/drools/reliability/infinispan/ReliabilityTestBasics.java [hunk 8]
```diff
@@ -187,7 +191,8 @@ protected void disposeSession() {
 
     protected KieSession getKieSession(String drl, PersistedSessionOption persistedSessionOption, Option... options) {
         OptionsFilter optionsFilter = new OptionsFilter(options);
-        KieBase kbase = new KieHelper().addContent(drl, ResourceType.DRL).build(optionsFilter.getKieBaseOptions());
+        //KieBase kbase = new KieHelper().addContent(drl, ResourceType.DRL).build(optionsFilter.getKieBaseOptions());
```
Comment (mariofusco) [inline drools-reliability/drools-reliability-infinispan/src/test/java/org/drools/reliability/infinispan/ReliabilityTestBasics.java]: Please remove this comment out line.

**Review by tkobayas — APPROVED** [review]

File: drools-core/src/main/java/org/drools/core/phreak/SynchronizedPropagationList.java (hunk 4 above)
Comment (nprentza) [reply drools-core/src/main/java/org/drools/core/phreak/SynchronizedPropagationList.java]: Yes, correct.

File: drools-core/src/main/java/org/drools/core/phreak/PropagationEntry.java (hunk 1 above)
Comment (nprentza) [reply drools-core/src/main/java/org/drools/core/phreak/PropagationEntry.java]: Done.

File: drools-core/src/main/java/org/drools/core/phreak/PropagationEntry.java (hunk 2 above)
Comment (nprentza) [reply drools-core/src/main/java/org/drools/core/phreak/PropagationEntry.java]: Done.

File: drools-core/src/main/java/org/drools/core/phreak/PropagationEntry.java (hunk 3 above)
Comment (nprentza) [reply drools-core/src/main/java/org/drools/core/phreak/PropagationEntry.java:231]: Ok.

File: drools-core/src/main/java/org/drools/core/phreak/SynchronizedPropagationList.java (hunk 5 above)
Comment (nprentza) [reply drools-core/src/main/java/org/drools/core/phreak/SynchronizedPropagationList.java]: Done.

File: drools-reliability/drools-reliability-core/src/main/java/org/drools/reliability/core/ReliableAgenda.java (hunk 6 above)
Comment (nprentza) [reply drools-reliability/drools-reliability-core/src/main/java/org/drools/reliability/core/ReliableAgenda.java]: Done.

File: drools-reliability/drools-reliability-core/src/main/java/org/drools/reliability/core/ReliableSessionInitializer.java (hunk 7 above)
Comment (nprentza) [reply drools-reliability/drools-reliability-core/src/main/java/org/drools/reliability/core/ReliableSessionInitializer.java]: Done.

File: drools-reliability/drools-reliability-infinispan/src/test/java/org/drools/reliability/infinispan/ReliabilityTestBasics.java (hunk 8 above)
Comment (nprentza) [reply drools-reliability/drools-reliability-infinispan/src/test/java/org/drools/reliability/infinispan/ReliabilityTestBasics.java]: Done.

**Comment by kie-ci4** [comment]
**(tests) - kogito-springboot-examples job** `#600` was: **UNSTABLE**
Possible explanation: This should be test failures


<details>
<summary><b>Reproducer</b></summary>


export BUILD_MVN_OPTS_CURRENT= 
build-chain build cross_pr -f 'https://raw.githubusercontent.com/kiegroup/drools/main/.ci/buildchain-config.yaml' -o 'bc' -p kiegroup/kogito-examples -u https://github.com/kiegroup/drools/pull/5239 --skipParallelCheckout

NOTE: To install the build-chain tool, please refer to https://github.com/kiegroup/github-action-build-chain#local-execution

</details>


Please look here: https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/drools/job/main/job/pullrequest/job/drools.tests.downstream.kogito-springboot-examples/600/display/redirect

**Test results:**
- PASSED: 125
- FAILED: 2

Those are the test failures: 
<details>
<summary><a href="https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/drools/job/main/job/pullrequest/job/drools.tests.downstream.kogito-springboot-examples/600/testReport/org.kie.kogito.examples.springboot/GrafanaDockerComposeIT/___/">org.kie.kogito.examples.springboot.GrafanaDockerComposeIT.(?)</a></summary>
org.testcontainers.containers.ContainerLaunchException: Local Docker Compose exited abnormally with code 18 whilst running command: --compatibility up -d<br/>
</details>
<details>
<summary><a href="https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/drools/job/main/job/pullrequest/job/drools.tests.downstream.kogito-springboot-examples/600/testReport/org.kie.kogito.springboot.outbox/OutboxIT/___/">org.kie.kogito.springboot.outbox.OutboxIT.(?)</a></summary>
org.testcontainers.containers.ContainerLaunchException: Local Docker Compose exited abnormally with code 18 whilst running command: --compatibility up -d kafka mongodb connect sidecar kogito<br/>
</details>

**Review by mariofusco — APPROVED** [review]

**Pull request merged**